### PR TITLE
[2036] Add page actions in form

### LINKED ISSUE #2036: Add the concept of page action in form
Continuation of issue https://github.com/eclipse-sirius/sirius-components/issues/1915.

In the same way as for groups, add to the view DSL `toolbarActions` for pages.

### PR DESCRIPTION
**Pull request by frouene**
Bug: https://github.com/eclipse-sirius/sirius-components/issues/2036

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -91,6 +91,7 @@ Introducing a generator of builders aimed to help creating view programmatically
 - https://github.com/eclipse-sirius/sirius-components/issues/1989[#1979] [diagram] Contribute a new way to render diagrams to evaluate an alternate layouting strategy
 - https://github.com/eclipse-sirius/sirius-components/issues/1996[#1996] [explorer] Add an Expand All menu item contribution to the Explorer View.
 - https://github.com/eclipse-sirius/sirius-components/issues/1966[#1966] [view] Add Selection Description in the View DSL
+- https://github.com/eclipse-sirius/sirius-components/issues/2036[#2036] [form] Add the concept of pageAction in form.
```
Comment (AxelRICHARD) [inline CHANGELOG.adoc]: Could tou please move this into 2023.8 section of the changelog?

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddToolbarActionEventHandler.java [hunk 2]
```diff
@@ -45,6 +46,7 @@
 @Service
 public class AddToolbarActionEventHandler implements IFormDescriptionEditorEventHandler {
 
+
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddToolbarActionEventHandler.java]: Could you please remove this superfluous line?

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddToolbarActionEventHandler.java [hunk 3]
```diff
@@ -99,12 +99,16 @@ private boolean addToolbarAction(IEditingContext editingContext, IFormDescriptio
...
-                ((GroupDescription) container).getToolbarActions().add(toolbarActionDescription);
+            var toolbarActionDescription = ViewFactory.eINSTANCE.createButtonDescription();
+            toolbarActionDescription.setName("ToolbarAction");
+            var toolbarActionDescriptionStyle = ViewFactory.eINSTANCE.createButtonDescriptionStyle();
+            toolbarActionDescription.setStyle(toolbarActionDescriptionStyle);
+            if (container instanceof GroupDescription groupDescription) {
+                groupDescription.getToolbarActions().add(toolbarActionDescription);
+                success = true;
+            }
+            if (container instanceof PageDescription pageDescription) {
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/AddToolbarActionEventHandler.java]: else if

File: packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveToolbarActionEventHandler.java [hunk 4]
```diff
@@ -108,17 +109,25 @@ private boolean moveToolbarAction(IEditingContext editingContext, IFormDescripti
...
+                            groupDescription.getToolbarActions().move(index, toolbarActionToMove);
                         } else {
-                            ((GroupDescription) container).getToolbarActions().add(index, toolbarActionToMove);
+                            groupDescription.getToolbarActions().add(index, toolbarActionToMove);
                         }
                         success = true;
-                    } catch (IndexOutOfBoundsException exception) {
-                        this.logger.warn(exception.getMessage(), exception);
                     }
+                    if (container instanceof PageDescription pageDescription) {
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/backend/sirius-components-collaborative-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/collaborative/formdescriptioneditors/handlers/MoveToolbarActionEventHandler.java]: else if

File: packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/components/FormDescriptionEditorPageComponent.java [hunk 5]
```diff
@@ -45,6 +61,16 @@ public Element render() {
         String label = this.getPageLabel(pageDescription, "Page");
         List<Element> childrenWidgets = new ArrayList<>();
 
+        pageDescription.getToolbarActions().forEach(viewToolbarActionDescription -> {
+            VariableManager childVariableManager = variableManager.createChild();
+            childVariableManager.put(VariableManager.SELF, viewToolbarActionDescription);
+            AbstractWidgetDescription toolbarActionDescription = this.converter.doSwitch(viewToolbarActionDescription);
+            if (toolbarActionDescription instanceof ButtonDescription) {
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/components/FormDescriptionEditorPageComponent.java]: instanceof ButtonDescription buttonDescription

File: packages/view/backend/sirius-components-view-edit/src/main/java/org/eclipse/sirius/components/view/provider/PageDescriptionItemProvider.java [hunk 6]
```diff
@@ -216,6 +218,8 @@ protected void collectNewChildDescriptors(Collection<Object> newChildDescriptors
         super.collectNewChildDescriptors(newChildDescriptors, object);
 
         newChildDescriptors.add(this.createChildParameter(ViewPackage.Literals.PAGE_DESCRIPTION__GROUPS, ViewFactory.eINSTANCE.createGroupDescription()));
+
+        newChildDescriptors.add(this.createChildParameter(ViewPackage.Literals.PAGE_DESCRIPTION__TOOLBAR_ACTIONS, ViewFactory.eINSTANCE.createButtonDescription()));
```
Comment (AxelRICHARD) [inline packages/view/backend/sirius-components-view-edit/src/main/java/org/eclipse/sirius/components/view/provider/PageDescriptionItemProvider.java]: Could you please take a look at `GroupDescriptionItemProvider#collectNewChildDescriptors` and apply the same modifications to `PageDescriptionItemProvider#collectNewChildDescriptors`?

**Review by AxelRICHARD — CHANGES_REQUESTED** [review]

File: packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/components/FormDescriptionEditorPageComponent.java [hunk 7]
```diff
@@ -32,9 +39,18 @@ public class FormDescriptionEditorPageComponent implements IComponent {
 
     private final FormDescriptionEditorPageComponentProps props;
 
+    private final Switch<AbstractWidgetDescription> converter;
+
```
Comment (AxelRICHARD) [inline packages/formdescriptioneditors/backend/sirius-components-formdescriptioneditors/src/main/java/org/eclipse/sirius/components/formdescriptioneditors/components/FormDescriptionEditorPageComponent.java:43]: Please remove this superfluous line

**Review by AxelRICHARD — DISMISSED** [review]

**Review by pcdavid — APPROVED** [review]

**Review by pcdavid — APPROVED** [review]

**Comment by pcdavid** [comment]
Rebased and tested locally. I'll merge as soon as the build is green.

**Pull request merged**